engine/table/src/main/java/io/deephaven/engine/table/impl/dataindex/DataIndexProviderImpl.java

```diff
@@ -0,0 +1,132 @@
+import io.deephaven.engine.table.impl.locations.TableLocation;
+import io.deephaven.engine.table.impl.perf.QueryPerformanceRecorder;
+import io.deephaven.engine.util.TableTools;
+import io.deephaven.util.annotations.InternalUseOnly;
+import io.deephaven.vector.ObjectVector;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.lang.ref.SoftReference;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class provides data indexes for merged tables. It is responsible for ensuring that the provided table accounts
+ * for the relative positions of individual table locations in the provided table of indices.
+ *
+ * <p>
+ * Much like the {@link GroupingProvider} implementations, this also attempts to defer any actual disk accesses until
+ * they are absolutely necessary.
+ *
+ * @implNote This is an experimental feature, it is likely to change.
+ */
+@InternalUseOnly
+public class DataIndexProviderImpl implements DataIndexProvider {
+    private final LinkedHashMap<TableLocation, LocationState> locations = new LinkedHashMap<>();
+
+    private static class LocationState {
+        private final TableLocation location;
+        private final long firstKey;
+
+        private Map<List<String>, SoftReference<Table>> dataIndices;
+
+        private LocationState(TableLocation location, long firstKey) {
+            this.location = location;
+            this.firstKey = firstKey;
+        }
+
+        @Nullable
+        private Table getDataIndex(@NotNull final String... keyColumns) {
+            // Sort the key columns so we can depend on the list as a lookup key.
+            // make a copy of the collection so we don't permute the parameter array.
+            final List<String> indexKey = Arrays.stream(keyColumns)
+                    .sorted(String::compareTo)
+                    .collect(Collectors.toList());
+
+            synchronized (LocationState.this) {
+                if (dataIndices != null) {
+                    final SoftReference<Table> tableRef = dataIndices.get(indexKey);
+                    if (tableRef != null) {
+                        final Table result = tableRef.get();
+                        if (result != null) {
+                            return result;
+                        } else {
+                            dataIndices.remove(indexKey);
+                        }
+                    }
+                }
+
+                Table indexTable = location.getDataIndex(keyColumns);
+                if (indexTable != null) {
+                    // Shift the indices onto our range.
+                    indexTable = indexTable.update("Index=Index.shift(" + firstKey + ")");
+
+                    if (dataIndices == null) {
+                        dataIndices = new HashMap<>();
+                    }
+                    dataIndices.put(indexKey, new SoftReference<>(indexTable));
+                }
+                return indexTable;
+            }
```

Comment:
I think we should make this lock smaller. This might be a use-case for having a persistent `DataIndexState` object linking an immutable `indexKey` and a mutable `volatile SoftReference`, maybe with a cached "canCompute" `boolean` to avoid asking for the same index more than once if the location can't provide it.

Having a re-usable object will allow us to lock at finer granularity if we need to concurrently build two different indices. Maybe use `KeyedObjectHashMap`.